cpp/src/stream_compaction/unique_count.cu

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "stream_compaction_common.cuh"
+#include "stream_compaction_common.hpp"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/detail/stream_compaction.hpp>
+#include <cudf/stream_compaction.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/count.h>
+#include <thrust/execution_policy.h>
+#include <thrust/logical.h>
+
+#include <cmath>
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace cudf {
+namespace detail {
+namespace {
+/**
+ * @brief A functor to be used along with device type_dispatcher to check if
+ * the row `index` of `column_device_view` is `NaN`.
+ */
+struct check_nan {
+  // Check if it's `NaN` for floating point type columns
+  template <typename T, std::enable_if_t<std::is_floating_point_v<T>>* = nullptr>
+  __device__ inline bool operator()(column_device_view const& input, size_type index)
+  {
+    return std::isnan(input.data<T>()[index]);
+  }
+  // Non-floating point type columns can never have `NaN`, so it will always return false.
+  template <typename T, std::enable_if_t<not std::is_floating_point_v<T>>* = nullptr>
+  __device__ inline bool operator()(column_device_view const&, size_type)
+  {
+    return false;
+  }
+};
```

Comment:
Nitpick: At some point, we should change this to:
```suggestion
  template <typename T>
  __device__ inline bool operator()(column_device_view const& input, size_type index)
  {
    if constexpr(std::is_floating_point_v<T>) {
      return std::isnan(input.data<T>()[index]);
    }
    else {
      return false;
    }
  }
```

Reply:
We might also drop the `inline`, since this is a function template.

Reply:
I noticed by doing so, build time significantly increased.

Reply:
@mythrocks I tried to encourage `if constexpr` over SFINAE in cases like this in the past and I believe that other reviewers disagreed about it being simpler/easier to read. I'm not sure what the exact case was. I would support `if constexpr` here, but this probably does need to be `inline`. At least, I see no reason _not_ to `inline`.

Reply:
(Either way, I'd prefer to keep PR momentum and not get too hung up here.)

Reply:
SFINAE -> `if constexpr`

The killer

Reply:
> I noticed by doing so, build time significantly increased.

That comes as a surprise. Let's hold off on making that change.

> this probably does need to be inline.

Unspecialized template, no? `inline` is superfluous, because it isn't subject to ODR?
Not important at all.

I'm resolving this suggestion.